lib/vrt/mapping.rb

```diff
@@ -50,14 +56,37 @@ def load_mappings
     # becomes
     #     {one: {'id': 'one', 'foo': 'bar'}, two: {'id': 'two', 'foo': 'baz'}}
     def key_by_id(mapping)
-      case mapping
-      when Array
+      if mapping.is_a?(Array) && mapping.first.is_a?(Hash) && mapping.first.key?('id')
         mapping.each_with_object({}) { |entry, acc| acc[entry['id'].to_sym] = key_by_id(entry) }
-      when Hash
+      elsif mapping.is_a?(Hash)
         mapping.each_with_object({}) { |(key, value), acc| acc[key] = key_by_id(value) }
       else
         mapping
       end
     end
+
+    def get_key(id_list:, mapping:, key:, default:)
+      # iterate through the id components, keeping track of where we are in the mapping file
+      # and the most specific mapped value found so far
+      best_guess = default
+      id_list.each do |id|
+        entry = mapping[id]
+        break unless entry # mapping file doesn't go this deep, return previous value
+        best_guess = merge_collections(best_guess, entry[key]) if entry[key]
+        # use the children mapping for the next iteration
+        mapping = entry['children'] || {}
+      end
+      best_guess
```

Comment:
I feel like there's a way to make this method a lot simpler but I can't quite crack it!